### PR TITLE
ARROW-3018: [Plasma] Remove Mersenne twister

### PR DESCRIPTION
**Pull request by pcmoritz**
This removes the C++ random ObjectID generator. Having a global random number generator is not a good idea and with an existing random number generator, it is very easy to write code like:

```
std::string random_string = // Create random string of length ObjectID::size()
auto object_id = ObjectID::from_binary(random_string);
```

### CODE REVIEW

File: cpp/src/plasma/common.cc [hunk 1]
```diff
@@ -29,6 +30,13 @@ namespace plasma {
 
 using arrow::Status;
 
+std::mt19937 RandomlySeededMersenneTwister() {
+  auto seed = static_cast<std::mt19937::result_type>(
```
Comment (pitrou) [inline cpp/src/plasma/common.cc]: I don't think you need to compute your own seed, just use `std::random_device` as shown in the example here: https://en.cppreference.com/w/cpp/numeric/random

Comment (pitrou) [reply cpp/src/plasma/common.cc]: Note that C++ doesn't allow use to correctly seed the MT engine, but for this purpose it probably doesn't matter much. See http://www.pcg-random.org/posts/cpp-seeding-surprises.html

**Review by pitrou — CHANGES_REQUESTED** [review]
Hi. I left two comments. Also, I think this warrants a test (which should be easily written on the Python side, using `subprocess`). Also you might want to protect the generator against forking (which you could test using `multiprocessing`).

File: cpp/src/plasma/common.cc (hunk 1 above)
Comment (pcmoritz) [reply cpp/src/plasma/common.cc]: Thanks for the feedback! Unfortunately std::random_device breaks the valgrind that ships with ubuntu 16.04 (it uses a processor instruction that is not implemented in that version of valgrind, but if you compile the latest valgrind it works).

So our choices are:
- Read randomness from /dev/urandom (might not be portable)
- Come up with some good way to seed ourselves (most likely doesn't give the best possible randomness)
- Use random_device, upgrade valgrind on Travis and document somewhere that a newer version of valgrind is needed

Let me know if you have thoughts on this!

Comment (pitrou) [reply cpp/src/plasma/common.cc]: Ah, right, I got this locally too. Then keep the current seed and add a comment about Valgrind?

**Comment by pcmoritz** [comment]
@pitrou Thanks, I fixed the things you pointed out now (the seed should now be strong and also it should work with older versions of valgrind).

I'm not sure what you mean with `you might want to protect the generator against forking` however, can you elaborate?

**Comment by pitrou** [comment]
What I mean is what happens when you do this:
```python
>>> import multiprocessing as mp
>>> from pyarrow import plasma
>>> def child(): print(plasma.ObjectID.from_random())
>>> 
>>> for i in range(4): mp.Process(target=child).start()
ObjectID(c0d241d281f0b269e46cf5948c2823c3263a41cf)
ObjectID(c0d241d281f0b269e46cf5948c2823c3263a41cf)
ObjectID(c0d241d281f0b269e46cf5948c2823c3263a41cf)
ObjectID(c0d241d281f0b269e46cf5948c2823c3263a41cf)
```

You want to re-seed or re-create your generator in each new child process...

**Comment by pcmoritz** [comment]
This is now ready to merge now. After some deliberation I decided it was better to get rid of the C++ random number generator altogether.

I don't think deprecation is necessary here since it is very easy to make a random object id but let me know if you think otherwise!

**Review by wesm — APPROVED** [review]
+1. Removing the `from_random` method seems OK to me. Merging, thanks @pcmoritz!

**Comment by robertnishihara** [comment]
Merging now, I believe the comments have been addressed.

**Pull request closed without merging**